# Feedback from previous weeks and other and hints

1. Be careful about your environment remembering variables. Make sure your code works in a new _clean_ environment. In Colab: `Runtime`->`restart Runtime`, in Anaconda's Jupyter: `Kernel`->`Restart`.
2. Graphs without labels (or units when appropriate) are not worth any point.
3. Do put in sufficient explanatory comments in your code.
4. Functions are very important. Do look up the video on the Safari O'Reilly ressource if you are still not clear on them !

For this week you can use these imports at the start of your programs:

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

We will use a new module `uncertainties`, which is **not** standard in the colab environment. You will have to first run:

In [ ]:
! pip install -q uncertainties

To install the module before you can import it:

In [ ]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions

# Introduction
In the practical classes PX2133/PX2233 and PX2338 (Obs tech), as well as your year 3/4 project, a lot of emphasis is placed on the determination and mathematical handling of errors.
The uncertainties module allows us to deal very easily with [error propagation](https://en.wikipedia.org/wiki/Propagation_of_uncertainty). For this sheet you should remind yourself about error bars in measurements and about propagation of uncertainties. Take an example from your lab handbook:

**Example 1**: If the length of a rectangle is $1.24\pm0.02 m$ and its breadth is $0.61\pm0.01 m$, what is its area and the error in the area? The following code snippet solves this problem in a few lines.

In [ ]:
L = uc.ufloat(1.24, 0.02)
W = uc.ufloat(0.61, 0.01)
print ('Area is:', L*W, 'm^2') # Do remember to add the units when printing!

Area is: 0.756+/-0.017 m^2


**Note**: For the area itself, it's fairly straightforward:

In [ ]:
1.24*0.61

0.7564

However, for the error bar on this number:

In [ ]:
0.02*0.01

0.0002

does not work. Instead, the [error progation formula](https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulae) gives:

In [ ]:
np.abs(1.24*0.61)*np.sqrt((0.02/1.24)**2+(0.01/0.61)**2)

0.01739540169125163

So the area is $0.756\pm0.017 m^2$. `uncertainties` obviously saves a lot of work, even for such a simple case. You can also take a look at the web site uncertainties hosted at https://pythonhosted.org/uncertainties/user_guide.html. In particular, [this section](https://pythonhosted.org/uncertainties/user_guide.html#access-to-the-uncertainty-and-to-the-nominal-value) shows some of the properties of a `ufloat` you can access directly.

**Example 2**: A reference object is $10.0\pm0.0001 m$ long, and makes a viewing angle of $0.62\pm0.02 rad$. How far is it?

In [ ]:
L = uc.ufloat(10.0, 0.0001)
theta = uc.ufloat(0.62,0.02)

Distance = (L/2)/um.tan(theta/2)

print ('Distance is:', Distance.nominal_value, 'm, with an error of:', Distance.std_dev)

Distance is: 15.609024890896208 m, with an error of: 0.537283338762715


Note the need to use "umath" functions (like `um.tan()` instead of `np.tan()`), and how to get the nominal value and the standard deviation of the uncertainties objects. To get nicer looking output, such as controlling the number of significant digits printed, you can use the information about formatting at https://docs.python.org/3/tutorial/inputoutput.html. In the exercises below you need to print the values to the screen. (Don’t forget units.)

# Exercises
This must be marked before you leave the lab. Mark weighting is in brackets.
**Save your work to GitHub after having run all cells with `Runtime` -> `Restart and run all`. And do not change the notebook's filename.** Do add comments to your code, you'll lose points if your code is hard to understand. Graphs without labels (or units when appropriate) are not worth any point.

## Exercise 0
[0] With some approximations, we have measured the mass of the following black-holes:
```
"35.6+/-3.9","30.6+/-3.7","63.1+/-3.2","23.2+/-9.8","13.6+/-4.5","35.7+/-6.8","13.7+/-6.0","7.7+/-2.4","20.5+/-4.0"
```
Compute for each (with error-bars) their lifetime due to Hawking radiation:
$$
t = \left(\frac{M}{M_{\odot}}\right)^3\,\times\,2.097\,\times\,10^{67} yr
$$

(this exercise is for demonstration purposes and won't be marked)

## Exercise 1
[2] An object is measured to travel a distance $x = 5.1 \pm 0.4 m$ during a time of $t = 0.4 \pm 0.1 s$. What is the average velocity and the error in the average velocity?

In [9]:
import uncertainties as uc

x = uc.ufloat(5.1, 0.4)
t = uc.ufloat(0.4,0.1)

vel = x/t

print(f'Average Velocity of the object is {vel.nominal_value:.3f} ms⁻¹ which has an error of ± {vel.std_dev:.3f} ms⁻¹')

Average Velocity of the object is 12.750 ms⁻¹ which has an error of ± 3.341 ms⁻¹


## Exercise 2
[2] An enterprising cow attempts to jump over the moon by jumping vertically into the air with initial speed $v_0=4.0\pm0.2 m/s$. After a time $t=0.60\pm0.06s$, the height of the cow is $h = v_0t-\frac{1}{2}g t^2 = 0.636 m$. What is the uncertainty in $h$? Take $g$ as exactly $9.81 ms^{-2}$.

In [26]:
t = uc.ufloat(0.60, 0.06)
v0 = uc.ufloat(4.0, 0.2)
g = 9.81

h = v0*t - 0.5*g*t**2

print(f'The uncertainty in h is {h.std_dev:.3f} m, The value of the height is {h.nominal_value:.3f}')

The uncertainty in h is 0.165 m, The value of the height is 0.634


## Exercise 3
[2] In an optics experiment the object distance $u$ is measured to be 20cm and the image distance $v$ is 10cm, both to an accuracy of 0.5cm. Find the focal length $f$ of the lens using the formula:

$$ \frac{1}{u}+\frac{1}{v}=\frac{1}{f}$$

In [25]:
u = uc.ufloat(20, 0.5)
v = uc.ufloat(10, 0.5)

f = 1/((1/u)+(1/v))

print(f'The focal length of the lens is {f.nominal_value:.3f} cm with an error of {f.std_dev:.3f}')

The focal length of the lens is 6.667 cm with an error of 0.229


## Exercise 4
[2] Two students each measure the refractive index of water. Jack measures a value of $1.33 \pm 0.03$ while Jill measures $1.28 \pm 0.02$. Are these values in agreement? *You do have to think a bit about this one...*

### I think the answers are in agreement:
As there is an overlap in answers at the value 1.30 within their error bars. I would say agreement however you would need more measurements to be more sure of the actual value

In [ ]:
#My answer is in the text box above

## Exercise 5
[2] The damped resonance frequency $\omega_{res}$ of an oscillating system is related to the (un-damped) natural angular frequency $\omega_0$ and the damping coefficient $\alpha$ by:

$$\omega_{res} = \sqrt{ \omega_0^2 - 2\alpha^2}$$

Find $f_0$ if the measured resonance frequency $f_{res}$ is $23.2\pm0.1 Hz$ and the measured damping coefficient is $19.5\pm0.5s^{-1}$.


In [27]:
import numpy as np
import uncertainties.umath as um
fRes = uc.ufloat(23.2, 0.1)
a = uc.ufloat(19.5, 0.5)

wRes = 2 * np.pi * fRes

w0 = um.sqrt((wRes**2) + 2*a**2)

f0 = w0/(2*np.pi)

print(f'f0 is {f0.nominal_value:.3f}Hz with an error of {f0.std_dev:.3f}Hz')

f0 is 23.612Hz with an error of 0.100Hz


## Exercise 6
[2] Suppose you have the following equation from one of your lab experiments:

$$f=\frac{c}{2}\sqrt{\frac{n_x^2}{L_x^2}+\frac{n_y^2}{L_y^2}+\frac{n_z^2}{L_z^2}}$$

where $f$ is the resonant frequency of sound waves in a box of sides $L_x$, $L_y$ and $L_z$ in length and the $n_x$ etc. are integers. $L_x = 10.2\pm0.2m$, $L_y = 5.2\pm0.3m$ and $L_z = 20.0\pm0.1 m$, while $c = 331.3 + T * 0.606 \,m\,s^{-1}$ is the temperature-dependent speed of sound, and the temperature $T$ is $23 \pm 1^\circ C$.
Calculate $f$ and the error in $f$ for the following values of $(nx,ny,nz)= (1,1,1), (1,1,2)$ and $(2,1,1)$.


In [35]:
lx = uc.ufloat(10.2, 0.2)
ly = uc.ufloat(5.2, 0.3)
lz = uc.ufloat(20.0, 0.1)

T = uc.ufloat(23,1)

c = 331.3 + T*0.606

def f(n):
  nx, ny, nz = n

  return (c/2)*(um.sqrt((nx**2/lx**2)+(ny**2/ly**2)+(nz**2/lz**2)))
values = [[1,1,1], [1,1,2], [2,1,1]]

print(f'For values {values[0]} for (nx,ny,nz) you get a values of f as {f(values[0]).nominal_value}Hz withan error of {f(values[0]).std_dev}Hz')
print(f'For values {values[1]} for (nx,ny,nz) you get a values of f as {f(values[1]).nominal_value}Hz withan error of {f(values[1]).std_dev}Hz')
print(f'For values {values[2]} for (nx,ny,nz) you get a values of f as {f(values[2]).nominal_value}Hz withan error of {f(values[2]).std_dev}Hz')

For values [1, 1, 1] for (nx,ny,nz) you get a values of f as 38.24745337649154 withan error of 1.6700605379580133
For values [1, 1, 2] for (nx,ny,nz) you get a values of f as 41.06516265028539 withan error of 1.5562781935426817
For values [2, 1, 1] for (nx,ny,nz) you get a values of f as 48.18791630742113 withan error of 1.4018271041233974


## Exercise 7
[4] The reflection coefficient $R_\parallel$ for parallel plane-polarised light reflected from a surface is given by the equation:

$$ R_\parallel = \frac{\tan^2(\theta_i - \theta_t)}{\tan^2(\theta_i + \theta_t)} $$

Calculate the error in $R_\parallel$ given measurements $\theta_i = (78 \pm 1)^\circ$ and $\theta_t = (40 \pm 1)^\circ$.

In [36]:
thetai = uc.ufloat(78, 1)
thetaT = uc.ufloat(40, 1)

rP = (um.tan(thetai - thetaT)**2)/(um.tan(thetai + thetaT)**2)

print(f'The error in reflection coefficient is {rP.nominal_value} with an error of {rP.std_dev}')

The error in reflectino coefficient is 0.003572137026589645 with an error of 0.06506774439998199


## Exercise 8
[4] Calculate and print to the screen the fractional uncertainty, as a percentage to one
significant figure, of the fluid flow discharge coefficient $C_d$ from the equation

$$
C_d = \frac{\dot{m}\sqrt{1-\left(\frac{d}{D}\right)^4}}{Kd^2F\sqrt{\rho\Delta P}}
$$

where

\begin{align*}
    C_d &= \text{discharge coefficient}&& \text{(no units)} \\
    \dot{m} &= \text{mass flow rate}&& = 0.13 \pm 0.01kg\,s^{-1} \\
    d &= \text{orifice diameter}&& = 11\pm 1 mm \\
    D &= \text{pipe diameter}&& = 71 \pm 1 mm \\
    \rho &= \text{fluid density}&& =1.01\pm0.01g\,cm^{-3} \\
    \Delta P &= \text{differential pressure}&& =156 \pm 7 Pa \\
    K &= \text{a constant parameter}&& =\text{constant (no units)} \\
    F &= \text{thermal expansion factor}&& =\text{constant (no units)}
\end{align*}


In [38]:
m = uc.ufloat(0.13, 0.01) #kg s-1
d_mm = uc.ufloat(11, 1) #mm
D_mm = uc.ufloat(71, 1) #mm
rho_gcm = uc.ufloat(1.01, 0.01) #g cm-3
dP = uc.ufloat(156, 7) #Pa
K = 20
F = 1e-6

d = d_mm/1000
D = D_mm/1000
rho = rho_gcm/1000

cd = (m*um.sqrt(1-((d/D)**4)))/(K*d**2*F*um.sqrt(rho*dP))

print(f'The fractional uncertainty of the discharge coefficient is {cd.std_dev/cd.nominal_value*100:.1f}%')

The fractional uncertainty of the discharge coefficient is 19.9%


## Exercise 9: Optional problem (not marked)
If you have time and want to try something interesting, do the following problem by plotting in 2D:
 - Draw an equilateral triangle with vertices and coordinates: vertex 1: $(p_1,q_1)$; vertex 2: $(p_2, q_2)$; vertex 3: $(p_3, q_3)$.
 - Place a dot at an arbitrary point $P = (x_0, y_0)$ within this triangle.
 - Find the next point by selecting randomly an integer $n = 1 , 2, $  or $3$ :
    1. If 1 , place a dot halfway between P and vertex 1.
    2. If 2 , place a dot halfway between P and vertex 2.
    3. If 3 , place a dot halfway between P and vertex 3.
 - Repeat the last two steps using the last dot as the new P.

Mathematically, the coordinates of successive points are given by the formulae

$$(x_{i+1},y_{i+1})=0.5[(x_i,y_i)+(p_n,q_n)]$$

and

$$n=int(1+3r_i),$$

where $r_i$ is a random number between 0 and 1 and where the $int()$ function outputs the closest integer smaller than or equal to the argument.

Try extending this to four vertices.